src/enzyme_ad/jax/Utils.h

```diff
@@ -306,6 +307,23 @@ bool canApplyNoNanPattern(bool allowOnFloatingPointMath, Type outTy, Type inTy);
 bool canApplyNoNanPattern(bool allowOnFloatingPointMath, Type outTy, Type inTy,
                           mlir::Operation *op);
 
+class GuaranteedResultKindBase {
+protected:
+  llvm::DenseMap<mlir::Value, bool> valueCache;
+  llvm::DenseMap<mlir::Operation *, bool> opCache;
+
+  virtual bool constantFloatCheck(DenseElementsAttr attr) = 0;
```

Comment:
Virtual functions will make this a lot slower, can you do the same template overload trick as in subreducebase/etc